### ISSUE TITLE
Add framework for solids

### ISSUE THREAD

**Issue by shivupa** [chemreps/chemreps#55]
In order to develop representations for solids, we need a framework to parse csf files (maybe some others like poscar in the future?)

 This requires creating a `cell` or `crystal` class that inherits the molecule class and parses a csf file.

**Comment by Ryoo** [comment]
Instead of opening a new issue thread, I'm just continuing on here.

I'm not very familiar with .csf files, but I'm currently working on .cif file for solids/crystals in unit cell. I parsed fractional coordinates of atoms in the unit cell into Cartesian coordinates for the library.

The major issue I'm encountering is that there are usually multiple (>1) formula in a unit cell, and often the border of the unit cell is clipping the molecules. This leaves us molecules broken into small portions. Hoping to stitch the fragments as a whole molecule, I also parsed the space groups of the unit cell, which may require me to make a whole different script to somehow link the space group and their symmetry operation.

**Comment by amandadumi** [comment]
Hm! I’ll take a look. Can this be reproduced with the cyclo_pentane.cif in #69 ?

**Comment by Ryoo** [comment]
Yeah, I'm looking up to compile all the symmetry operations and place them with all the space group. There are 230 distinct space groups (incl. chiral ones), so it may take some time. Hopefully I can just import it and use it on the parser once it is done.

**Comment by shivupa** [comment]
I'm not sure I understand the issue here. Even if the molecule is clipped the periodic image of the clipped part is still present correct?

**Comment by Ryoo** [comment]
Yeah, so the script is parsing single molecule in a unit cell. If I give it say propane C1 - C2 - C3, and the unit cell clips somewhere on C2-C3, I get C3 and C1-C2. Computer will consider C3 and C1 having a long bond instead of clipped molecule. 

Of course in this case the problem can be solved if I translate fragments in unit cell length, but it can be more complicated if we look at much larger molecules going 3 dimensions and unit cell with like a mirror symmetry. 

Maybe it shouldn't be much of problem because .cif file also contain bond, angle, and distortion information. Or we could use such information to recreate non-fragmented molecule too.

**Comment by shivupa** [comment]
Correct me if I am wrong, but I think the functions that need distances need to account for the periodicity by being modified to check which of the periodic images are the closest. 

For example consider a cubic cell where a=b=c=1 and alpha=beta=gamma=90, and atom_1 is at [0.1,0.1,0.1] and atom_2 is at [0.9,0.9,0.9]. The shortest distance is a periodic image. 
The code below shows that.
```
import numpy as np


a = 1.0
b = 1.0
c = 1.0

alpha = np.pi / 2.0  # 90
beta = np.pi / 2.0  # 90
gamma = np.pi / 2.0  # 90

omega = (a * b * c) * np.sqrt(
    1
    - (np.cos(alpha) ** 2)
    - (np.cos(beta) ** 2)
    - (np.cos(gamma) ** 2)
    + (2 * np.cos(alpha) * np.cos(beta) * np.cos(gamma))
)

uvw_1 = np.array([0.1, 0.1, 0.1])
uvw_2 = np.array([0.9, 0.9, 0.9])

# in same cell
M = np.array(
    [
        [a, b * np.cos(gamma), c * np.cos(beta)],
        [
            0,
            b * np.sin(gamma),
            c * (np.cos(alpha) - (np.cos(beta) * np.cos(gamma))) / (np.sin(gamma)),
        ],
        [0, 0, omega / (a * b * np.sin(gamma))],
    ]
)
xyz_1 = M @ uvw_1
xyz_2 = M @ uvw_2
print(np.linalg.norm(xyz_2 - xyz_1))

# in different cells


def f(p, q, r):
    shifted_uvw_2 = uvw_2 + np.array([p, q, r])
    M = np.array(
        [
            [a, b * np.cos(gamma), c * np.cos(beta)],
            [
                0,
                b * np.sin(gamma),
                c * (np.cos(alpha) - (np.cos(beta) * np.cos(gamma))) / (np.sin(gamma)),
            ],
            [0, 0, omega / (a * b * np.sin(gamma))],
        ]
    )
    xyz_1 = M @ uvw_1
    xyz_2 = M @ shifted_uvw_2
    return np.linalg.norm(xyz_2 - xyz_1)


# https://stackoverflow.com/a/46839863
import itertools

combinations = list(itertools.product(*[[-1, 0, 1]] * 3))
values = []
for combi in combinations:
    values.append((combi, f(*combi)))
best = [(c, v) for c, v in values if v == min([v for c, v in values])]

print(best)

```

I vote in #69 to keep `self.xyz` to all be in the same unit cell since this is unambiguous, and to keep this in mind when bonds, angles and torsions are calculated.

**Comment by Ryoo** [comment]
I was not aware that the distance function considers unit cell and the periodicity. Having that accounted in the functions like distance and angle sounds reasonable to me. 

Sorry, but I'm not fully understanding what you mean by 'all be in the same unit cell.' Are you just planning to create arbitrary periodic condition that are all in same size ? All the crystals contains their unique unit cell, and those values can be parsed from .cif file. #69 translates fractional coordinates respect to the unit cell into Cartesian coordinates, but their unit cell remain the same - I don't think we can force the same arbitrary periodic condition may be difficult considering that we cannot change how they are stacked in the crystals.

**Comment by shivupa** [comment]
> I was not aware that the distance function considers unit cell and the periodicity. Having that accounted in the functions like distance and angle sounds reasonable to me.

They don't yet, but they will need to be modified to take this into account. (not necessarily by you and not in PR #69).

> Sorry, but I'm not fully understanding what you mean by 'all be in the same unit cell.' Are you just planning to create arbitrary periodic condition that are all in same size ? All the crystals contains their unique unit cell, and those values can be parsed from .cif file. #69 translates fractional coordinates respect to the unit cell into Cartesian coordinates, but their unit cell remain the same - I don't think we can force the same arbitrary periodic condition may be difficult considering that we cannot change how they are stacked in the crystals.

Yeah sorry my wording here was a bit unclear. What I mean is when parsing the cif the initial self.xyz should be assigned using the fractional coordinates as is done in #69. I was saying we don't need to worry about changing them if they fragment a molecule, and #69 can move forward.

**Comment by Ryoo** [comment]
Okay! Yeah, and I can look into doing the distance modification portion too. Let me know if anyone got a good starting point for it.
